File: Ruby/Veterinaria/Veterinaria.rb
module Veterinaria
    class Mascota
        def initialize nombre, edad
            @nombre = nombre
            @edad = edad
        end

        def comunicarse
            raise NotImplementedError
        end

        def presentarse
            "Me llamo " + @nombre + " soy un "
        end
    end

    class Perro < Mascota
        def initialize tamanho, nombre, edad
            @tamanho = tamanho
            super nombre, edad
        end

        def perseguirLaCola
            "Mer persigo la cola"
        end

        def comunicarse
            "guau, guau"
        end

        def presentarse
            super + "perro"
        end
    end

    class Gato < Mascota
        def aranhar
            "Te aranho"
        end

        def comunicarse
            "miau, miau"
        end

        def presentarse
            super + "gato"
        end
    end

    class Ave < Mascota
        def comunicarse
            "pio, pio"
        end

        def presentarse
            super + "ave"
        end
    end

    class Reptil < Mascota
        def comunicarse
            "pss.. pss"
        end

        def presentarse
            super + "reptil"
        end
    end

    class Cliente
        def initialize nombre
            @nombre = nombre
            @mascotas = Array.new
        end

        def agregarMascota mascota
            @mascotas += [mascota]
        end

        def reporteMascotas
            mensaje = ""
            @mascotas.each do |mascota|
                mensaje += mascota.presentarse() + "\n"
            end
            return mensaje
        end
    end
end

File: Java/Veterinaria/src/veterinaria/Veterinaria.java
package veterinaria;

public class Veterinaria {

    public static void main(String[] args) {
        Perro perro = new Perro("Chico", "Rambo", 1);
        System.out.println(perro.presentarse());
        System.out.println(perro.comunicarse());
        
        /* Una clase abstracta no se puede instanciar
        Mascota mascota = new Mascota("Fantasma", 12);
        System.out.println(mascota.comunicarse());
        */
        
        Gato gato = new Gato("Lucifer", 3);
        System.out.println(gato.presentarse());
        System.out.println(gato.aranhar());
        // Polimorfismo (poli muchas, fismo formas): muchas formas.
        System.out.println(gato.comunicarse());
        
        Gato gato2 = new Gato("Gandalf", 7);
        Cliente cliente = new Cliente();
        cliente.agregarMascota(gato);
        cliente.agregarMascota(perro);
        cliente.agregarMascota(gato2);
        //System.out.println(cliente.reporteMascotas());
        Ave ave = new Ave("Piolin", 1);
        cliente.agregarMascota(ave);
        
        Reptil reptil = new Reptil("Juancho", 1);
        cliente.agregarMascota(reptil);
        System.out.println(cliente.reporteMascotas());
    }
}
